pkg/provisioner/ironic/raid.go

```diff
@@ -156,28 +163,24 @@ func buildTargetSoftwareRAIDCfg(volumes []metal3v1alpha1.SoftwareRAIDVolume) (lo
 
 // BuildRAIDCleanSteps build the clean steps for RAID configuration from BaremetalHost spec
 func BuildRAIDCleanSteps(raidInterface string, target *metal3v1alpha1.RAIDConfig, actual *metal3v1alpha1.RAIDConfig) (cleanSteps []nodes.CleanStep, err error) {
-	err = checkRAIDConfigure(raidInterface, target)
+	_, err = checkRAIDInterface(raidInterface, target, actual)
 	if err != nil {
 		return nil, err
 	}
 
-	// No RAID
-	if raidInterface == noRAIDInterface {
-		return
-	}
-
 	// Software RAID
-	if raidInterface == softwareRAIDInterface {
+	if target != nil && target.SoftwareRAIDVolumes != nil {
 		// Ignore HardwareRAIDVolumes
-		if target != nil {
-			target.HardwareRAIDVolumes = nil
-		}
+		target.HardwareRAIDVolumes = nil
 		if actual != nil {
 			actual.HardwareRAIDVolumes = nil
 		}
 		if reflect.DeepEqual(target, actual) {
 			return
 		}
+		if len(target.SoftwareRAIDVolumes) == 0 && (actual == nil || actual.SoftwareRAIDVolumes == nil || len(actual.SoftwareRAIDVolumes) == 0) {
```

Comment:
`len(<slice>) == 0` is always true when `<slice> == nil` is true, so there's no need to check both.